File: java-sdk/src/main/java/io/fyno/core_java/RequestHandler.java
package io.fyno.core_java;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkInfo;
import android.os.Build;
import android.util.Log;
import io.fyno.core_java.utils.FynoContextCreator;
import io.fyno.core_java.utils.Logger;
import org.json.JSONObject;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.LinkedList;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

public class RequestHandler {
    private static final String TAG = "FynoRequest";
    static FynoContextCreator contextCreator = new FynoContextCreator();
    FynoCore core = new FynoCore();
    private static final int TIMEOUT = 3000;
    private static final long MAX_BACKOFF_DELAY = 60000;
    private static final int MAX_RETRIES = 3;

    // Queue for offline requests
    private static final LinkedList<Request> offlineRequestsQueue = new LinkedList<>();

    // Network callback to listen for connectivity changes
    private ConnectivityManager.NetworkCallback networkCallback = null;

    private static final ExecutorService requestProcessor = Executors.newFixedThreadPool(1);

    private static final AtomicInteger retries = new AtomicInteger(0);

    public static class Request {
        private final String url;
        private final JSONObject postData;
        private final String method;

        public Request(String url, JSONObject postData, String method) {
            this.url = url;
            this.postData = postData;
            this.method = method;
        }
    }

    public static void requestPOST(String r_url, JSONObject postDataParams, String method) {
        Request request = new Request(r_url, postDataParams, method);
        if (contextCreator.isInitialized()) {
            if (isNetworkConnected(contextCreator.context)) {
                processOfflineRequests();
                sendRequest(request);
            } else {
                offlineRequestsQueue.add(request);
            }
        } else {
            sendRequest(request);
        }
    }

//    private void sendRequest(Request request) {
//        Logger.i(TAG, "requestPOST: Started processing " + request.url);
//        requestProcessor.execute(() -> handleRetries(request));
//    }

    private static boolean isNetworkConnected(Context context) {
        ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            NetworkCapabilities capabilities = connectivityManager.getNetworkCapabilities(connectivityManager.getActiveNetwork());
            return capabilities != null && (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) || capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI));
        } else {
            // Fallback for older Android versions
            @SuppressWarnings("deprecation")
            NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
            return networkInfo != null && networkInfo.isConnected();
        }
    }

    private static void handleRetries(Request request) {
        int retryCount = retries.get();
        if (retryCount < MAX_RETRIES) {
            try {
                doRequest(request);
            } catch (Exception e) {
                Logger.i(TAG, "Request failed: " + e.getMessage());
                long delayMillis = calculateDelay(retryCount);
                try {
                    Thread.sleep(delayMillis);
                } catch (InterruptedException ex) {
                    Logger.e(TAG, "Sleep interrupted: " + ex.getMessage());
                }
                retries.incrementAndGet();
                handleRetries(request);
            }
        } else {
            Logger.w(TAG, "Max retries reached for request: " + request.url);
            offlineRequestsQueue.add(request);
        }
    }

    private static long calculateDelay(int retryCount) {
        return Math.min((long) Math.pow(2, retryCount), MAX_BACKOFF_DELAY);
    }

    private static void doRequest(Request request) throws Exception {
        HttpURLConnection conn = null;
        try {
            URL url = new URL(request.url);
            conn = (HttpURLConnection) url.openConnection();
            conn.setRequestProperty("Content-Type", "application/json");
            if (contextCreator.isInitialized()) {
                conn.setRequestProperty("Authorization", "Bearer " + FynoUser.getApi());
            }
            conn.setReadTimeout(TIMEOUT);
            conn.setConnectTimeout(TIMEOUT);
            conn.setRequestMethod(request.method);
            conn.setDoInput(true);
            conn.setDoOutput(true);

            if (request.postData != null) {
                OutputStream os = conn.getOutputStream();
                OutputStreamWriter writer = new OutputStreamWriter(os, "UTF-8");
                writer.write(request.postData.toString());
                writer.flush();
                writer.close();
                os.close();
            }

            int responseCode = conn.getResponseCode();

            Logger.d("RequestPost", "requestPOST method = " + request.method + " url = " + request.url + ": " + conn.getResponseMessage());

            if (responseCode >= 200 && responseCode < 300) {
                Logger.i("RequestPost", "requestPOST: " + conn.getResponseMessage());
            } else if (responseCode >= 400 && responseCode < 500) {
                Logger.i(TAG, "Request failed with response code: " + responseCode);
            } else {
                Logger.i(TAG, "Request failed with response code: " + responseCode);
                throw new Exception("Request failed with response code: " + responseCode);
            }
        } finally {
            if (conn != null) {
                conn.getInputStream().close();
                conn.disconnect();
            }
        }
    }

    static void processOfflineRequests() {
        while (!offlineRequestsQueue.isEmpty()) {
            Request request = offlineRequestsQueue.poll();
            sendRequest(request);
        }
    }
    private static void sendRequest(final Request request) {
        final CountDownLatch latch = new CountDownLatch(1);
        requestProcessor.execute(() -> {
            handleRetries(request);
            latch.countDown();
        });
        try {
            latch.await();
        } catch (InterruptedException e) {
            Logger.e(TAG, "Thread interrupted while sending the request: " + e.getMessage());
        }
    }
}

File: fyno-kotlin-core/src/main/java/io/fyno/core/RequestHandler.kt
package io.fyno.core

import android.annotation.SuppressLint
import android.content.Context
import android.database.Cursor
import android.util.Log
import io.fyno.core.FynoCore.Companion.TAG
import io.fyno.core.helpers.SQLDataHelper
import io.fyno.core.utils.FynoContextCreator
import io.fyno.core.utils.Logger
import kotlinx.coroutines.*
import org.json.JSONObject
import java.io.BufferedReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import kotlin.math.pow
import kotlin.properties.Delegates

object RequestHandler {

    private const val TIMEOUT = 6000
    private const val MAX_BACKOFF_DELAY: Long = 60000
    private const val MAX_RETRIES = 1
    private lateinit var sqlDataHelper: SQLDataHelper

    interface NetworkCallback {
        fun onSuccess(response: String)
        fun onError(error: String)
    }

    data class Request(val url: String?, val postData: JSONObject?, val method: String = "POST")

    private suspend fun handleRetries(
        request: Request,
        id: Int? = 0,
        context: Context? = null
    ): Boolean {
        return withContext(Dispatchers.IO){
            var retries = 0
            while (retries < MAX_RETRIES) {
                try {
                    Logger.d("RequestHandler", "handleRetries: Retry attempt $retries for ${request.url}")
                    val success =  async { doRequestAsync(request, id, context) }.await()
                    Logger.d("RequestHandler", "handleRetries: Request started: ${request.url}, retry count = $retries, res -> $success")
                    if (success) return@withContext true
                } catch (e: Exception) {
                    Logger.d("RequestHandler", "handleRetries: Request failed: ${e.message}")
                    val delayMillis = calculateDelay(retries)
                    delay(delayMillis)
                    retries++
                }
            }
            Logger.w("RequestHandler", "handleRetries: Max retries reached for request: ${request.url}")
            return@withContext false
        }
    }

    private fun calculateDelay(retryCount: Int): Long {
        return minOf(4.0.pow(retryCount.toDouble()).toLong() * 1000, MAX_BACKOFF_DELAY)
    }

    private suspend fun doRequestAsync(request: Request, id: Int? = 0, context: Context? = null): Boolean {
        val mergeRegex = Regex(".*/track/(test|live)/profile/.*/merge/(.*)")
        return withContext(Dispatchers.IO) {
            val url = URL(request.url)
            val conn: HttpURLConnection = url.openConnection() as HttpURLConnection
            if (url.path.matches(Regex(".*/track/(test|live)/profile$"))) {
                request.postData?.getString("distinct_id")?.let { distinctId ->
                    val token = async { JWTRequestHandler().fetchAndSetJWTToken(distinctId) }.await()
                    if (token.isNullOrEmpty()) {
                        Logger.w(TAG,"JWT token fetch failed");
                        return@withContext false
                    }
                }
            }
            Logger.i(TAG, "doRequestAsync: request started for $url")
            conn.setRequestProperties()
            conn.readTimeout = TIMEOUT
            conn.connectTimeout = TIMEOUT
            conn.requestMethod = request.method
            conn.doInput = true
            conn.doOutput = true

            request.postData?.let { postData ->
                conn.outputStream.use { os ->
                    OutputStreamWriter(os, "UTF-8").use { writer ->
                        writer.write(postData.toString())
                    }
                }
            }

            val responseCode = conn.responseCode
            Logger.i(TAG, "doRequestAsync: Request finished with code : $responseCode")
            when (responseCode) {
                in 200..299 -> {
                    if (url.path.matches(mergeRegex)) {
                        mergeRegex.find(url.path)?.groups?.get(2)?.value?.let {
                            val token = async { JWTRequestHandler().fetchAndSetJWTToken(it) }.await()
                            if (token.isNullOrEmpty()) {
                                Logger.w(TAG,"JWT token fetch failed");
                                return@withContext false
                            }
                        }
                    }
                    if (isCallBackRequest(request.url)) {
                        deleteCBRequestFromDb(id, context?.applicationContext)
                        conn.disconnect()
                        return@withContext true
                    }
                    FynoContextCreator.sqlDataHelper.deleteRequestByID(id, "requests")
                    conn.disconnect()
                    true
                }
                in 400..499 -> {
                    if(responseCode == 404) {
                        conn.disconnect()
                        return@withContext false
                    }
                    val inputStream = conn.errorStream
                    val response = inputStream.bufferedReader().use(BufferedReader::readText)
                    val jsonResponse = JSONObject(response)
                    val message = jsonResponse.getString("_message")
                    if (responseCode == 401) {
                        if (message == "jwt_expired") {
                            async { JWTRequestHandler().fetchAndSetJWTToken(FynoUser.getIdentity()) }.await()
                            conn.disconnect()
                            async { doRequestAsync(request) }.await()
                        }
                    }
                    Logger.i(TAG, "doRequestAsync: Request failed with response code: $responseCode")
                    if (isCallBackRequest(request.url)) {
                        deleteCBRequestFromDb(id, context)
                        conn.disconnect()
                        return@withContext true
                    }
                    FynoContextCreator.sqlDataHelper.deleteRequestByID(id, "requests")
                    Logger.w(TAG, "doRequestAsync: Request failed: $message")
                    conn.disconnect()
                    true
                }
                else -> {
                    Logger.e(
                        TAG,
                        "doRequestAsync: Request failed with response code: $responseCode",
                        Exception("doRequestAsync: Request failed with response code: $responseCode")
                    )
                    conn.disconnect()
                    false
                }
            }
        }
    }

    private fun HttpURLConnection.setRequestProperties() {
        this.setRequestProperty("Content-Type", "application/json")
        if (FynoContextCreator.isInitialized()) {
            this.setRequestProperty("x-fn-app-id", FynoContextCreator.getContext()?.packageName)
            this.setRequestProperty("integration", FynoUser.getFynoIntegration())
            this.setRequestProperty("verify_token",FynoUser.getJWTToken())
        }
    }

    @SuppressLint("Range")
    suspend fun processDbRequests(caller: String? = "") {
        withContext(Dispatchers.IO) {
            async {
                var cursor: Cursor? = null
                var req_id by Delegates.notNull<Int>()
                try {
                    while (true) {
                        if (cursor != null && !cursor.isClosed) cursor.close()
                        cursor = FynoContextCreator.sqlDataHelper.getNextRequest("requests")
                        if (cursor.moveToNext()) {
                            val url =
                                cursor.getString(cursor.getColumnIndex(SQLDataHelper.COLUMN_URL))
                            val postDataStr =
                                cursor.getString(cursor.getColumnIndex(SQLDataHelper.COLUMN_POST_DATA))
                            val postData =
                                if (postDataStr != null) JSONObject(postDataStr) else null
                            val method =
                                cursor.getString(cursor.getColumnIndex(SQLDataHelper.COLUMN_METHOD))
                            val id = cursor.getInt(cursor.getColumnIndex(SQLDataHelper.COLUMN_ID))
                            val lastProcessedTimeMillis =
                                cursor.getLong(cursor.getColumnIndex(SQLDataHelper.COLUMN_LAST_PROCESSED_AT))
                            req_id = id
                            val timeDifference =
                                System.currentTimeMillis() - lastProcessedTimeMillis

                            if (caller != "requestPOST" && timeDifference < 2000) {
                                cursor.close()
                                continue
                            }
                            cursor.close()

                            FynoContextCreator.sqlDataHelper.updateStatusAndLastProcessedTime(
                                id,
                                "requests",
                                "processing"
                            )

                            val request = Request(url, postData, method)
                            val success = async { handleRetries(request, id) }.await();
                            if (success) {
                                FynoContextCreator.sqlDataHelper.deleteRequestByID(id, "requests")
                            } else {
                                FynoContextCreator.sqlDataHelper.updateStatusAndLastProcessedTime(
                                    id,
                                    "requests",
                                    "not_processed"
                                )
                                break;
                            }
                        } else {
                            break
                        }
                    }
                } catch (e: Exception) {
                    Logger.w(TAG, "Unable to process the request")
                    req_id.let {
                        FynoContextCreator.sqlDataHelper.updateStatusAndLastProcessedTime(
                            req_id,
                            "requests",
                            "not_processed"
                        )
                    }
                } finally {
                    cursor?.close()
                }
            }.await()
        }
    }

//    @SuppressLint("Range")
//    suspend fun processCBRequests(context: Context?, caller: String? = "") {
//        var cursor: Cursor? = null
//        var req_id by Delegates.notNull<Int>()
//        try {
//            while (true) {
//                if(cursor != null && !cursor.isClosed) cursor.close()
//                cursor = SQLDataHelper(context).getNextRequest("callbacks")
//                if (cursor.moveToNext()) {
//                    val url = cursor.getString(cursor.getColumnIndex(SQLDataHelper.COLUMN_URL))
//                    val postDataStr =
//                        cursor.getString(cursor.getColumnIndex(SQLDataHelper.COLUMN_POST_DATA))
//                    val postData = if (postDataStr != null) JSONObject(postDataStr) else null
//                    val method =
//                        cursor.getString(cursor.getColumnIndex(SQLDataHelper.COLUMN_METHOD))
//                    val id = cursor.getInt(cursor.getColumnIndex(SQLDataHelper.COLUMN_ID))
//                    req_id = id
//                    val lastProcessedTimeMillis =
//                        cursor.getLong(cursor.getColumnIndex(SQLDataHelper.COLUMN_LAST_PROCESSED_AT))
//                    val timeDifference = System.currentTimeMillis() - lastProcessedTimeMillis
//
//                    if (caller != "requestPOST" && timeDifference < 2000) {
//                        cursor.close()
//                        continue
//                    }
//                    cursor.close()
//
//                    SQLDataHelper(context).updateStatusAndLastProcessedTime(
//                        id,
//                        "callbacks",
//                        "processing"
//                    )
//
//                    val request = Request(url, postData, method)
//                    handleRetries(request, id, context)
//                } else {
//                    cursor.close()
//                    break
//                }
//            }
//        } catch (e: Exception) {
//            Logger.w(TAG, "Unable to process the request")
//            req_id.let {
//                SQLDataHelper(context).updateStatusAndLastProcessedTime(
//                    req_id,
//                    "callbacks",
//                    "not_processed"
//                )
//            }
//        } finally {
//            cursor?.close()
//        }
//    }

    @SuppressLint("Range")
    suspend fun processCBRequests(context: Context?, caller: String? = "") {
        withContext(Dispatchers.IO) {
            async {
                var cursor: Cursor? = null
                var req_id by Delegates.notNull<Int>()
                val db = SQLDataHelper(context)
                try {
                    while (true) {
                        if (cursor != null && !cursor.isClosed) cursor.close()
                        cursor = db.getNextRequest("callbacks")
                        if (cursor.moveToNext()) {
                            val url =
                                cursor.getString(cursor.getColumnIndex(SQLDataHelper.COLUMN_URL))
                            val postDataStr =
                                cursor.getString(cursor.getColumnIndex(SQLDataHelper.COLUMN_POST_DATA))
                            val postData =
                                if (postDataStr != null) JSONObject(postDataStr) else null
                            val method =
                                cursor.getString(cursor.getColumnIndex(SQLDataHelper.COLUMN_METHOD))
                            val id = cursor.getInt(cursor.getColumnIndex(SQLDataHelper.COLUMN_ID))
                            val lastProcessedTimeMillis =
                                cursor.getLong(cursor.getColumnIndex(SQLDataHelper.COLUMN_LAST_PROCESSED_AT))
                            req_id = id
                            val timeDifference =
                                System.currentTimeMillis() - lastProcessedTimeMillis

                            if (caller != "requestPOST" && timeDifference < 2000) {
                                cursor.close()
                                continue
                            }
                            cursor.close()

                            db.updateStatusAndLastProcessedTime(
                                id,
                                "callbacks",
                                "processing"
                            )

                            val request = Request(url, postData, method)
                            val success = async { handleRetries(request, id, context) }.await();
                            if (success) {
                                db.deleteRequestByID(id, "callbacks")
                            } else {
                                db.updateStatusAndLastProcessedTime(
                                    id,
                                    "callbacks",
                                    "not_processed"
                                )
                                db.close()
                                break;
                            }
                        } else {
                            db.close()
                            break
                        }
                    }
                } catch (e: Exception) {
                    Logger.w(TAG, "Unable to process the request")
                    req_id.let {
                        db.updateStatusAndLastProcessedTime(
                            req_id,
                            "callbacks",
                            "not_processed"
                        )
                    }
                } finally {
                    cursor?.close()
                    db.close()
                }
            }.await()
        }
    }

    suspend fun requestPOST(
        r_url: String?,
        postDataParams: JSONObject?,
        method: String = "POST",
        context: Context? = null
    ) {
        try {
            Logger.i(TAG, "requestPOST: Sending POST request")
            val request = Request(r_url, postDataParams, method)
            if (FynoContextCreator.isInitialized()) {
                if (isCallBackRequest(r_url)) {
                    saveCBRequestToDb(request, context)
                    processCBRequests(context, "requestPost")
                } else {
                    saveRequestToDb(request)
                    processDbRequests("requestPost")
                }
            } else {
                if (isCallBackRequest(r_url)) {
                    saveCBRequestToDb(request, context)
                    processCBRequests(context, "requestPost")
                }
            }
        } catch (e: Exception) {
            Logger.w(TAG, "requestPost: Failed to send request - ${e.stackTrace}")
            Logger.e(TAG, "requestPOST: ${e.message}",e )
        }
    }

    suspend fun requestPOSTWithCallbacks(
        r_url: String?,
        postDataParams: JSONObject?,
        method: String = "POST",
        context: Context? = null
    ) {
        try {
            Logger.i(TAG, "requestPOSTWithCallbacks: Sending POST request with callbacks")
            val request = Request(r_url, postDataParams, method)
            if (FynoContextCreator.isInitialized()) {
                if (isCallBackRequest(r_url)) {
                    saveCBRequestToDb(request, context)
                    processCBRequests(context, "requestPOST")
                } else {
                    saveRequestToDb(request)
                    processDbRequests("requestPOST")
                }
            } else {
                if (isCallBackRequest(r_url)) {
                    saveCBRequestToDb(request, context)
                    processCBRequests(context, "requestPOST")
                }
            }
        } catch (e: Exception) {
            Logger.w(TAG, "requestPOSTWithCallbacks: Failed to send request - ${e.message}")
        }
    }

    private fun isCallBackRequest(url: String?): Boolean {
        return !url.isNullOrEmpty() && (url.contains("callback.fyno.io") || url.contains("callback.dev.fyno.io"))
    }

    // Function to save requests to SQLite database
    private fun saveRequestToDb(request: Request, id:Int? = 0) {
        Logger.i(TAG, "saveRequestToDb: Saving request to db with url: ${request.url}")
        FynoContextCreator.sqlDataHelper.insertRequest(request, "requests", id)
    }

    // Function to save CB requests to SQLite database
    private fun saveCBRequestToDb(request: Request, context: Context?) {
        sqlDataHelper = SQLDataHelper(context)
        sqlDataHelper.insertRequest(request, "callbacks")
        sqlDataHelper.close()
    }

    private fun deleteCBRequestFromDb(id: Int?, context: Context?) {
        sqlDataHelper = SQLDataHelper(context?.applicationContext)
        sqlDataHelper.deleteRequestByID(id, "callbacks")
        sqlDataHelper.close()
    }
}
